### PR TITLE
feat: return `authorizationList` for EIP-7702 transactions in `/transactions/:tx_hash` response

### PR DESCRIPTION
**Pull request by AllFi**
Related to #10774

## Motivation

This PR introduces initial support for EIP-7702 transactions. It covers the transaction view part from #10774.

## Changelog
* Add the `signed_authorizations` table to store authorization tuples from EIP-7702 transactions.
* Enrich authorization tuples with the `authority` address recovered from the signature.
* Add `authorizationList` (including recovered authorities) to the `/transactions/:tx_hash` response.
* Add `set_code_transaction` tx type for EIP-7702 transactions.

## Checklist for your Pull Request (PR)

- [ ] If I added new functionality, I added tests covering it.
- [ ] If I fixed a bug, I added a regression test to prevent the bug from silently reappearing again.
- [ ] I checked whether I should update the docs and did so by submitting a PR to [docs repository](https://github.com/blockscout/docs).
- [ ] If I added/changed/removed ENV var, I submitted a PR to [docs repository](https://github.com/blockscout/docs) to update the list of [env vars](https://github.com/blockscout/docs/blob/master/for-developers/information-and-settings/env-variables.md) and I updated the version to `master` in the Version column. If I removed variable, I added it to [Deprecated ENV Variables](https://github.com/blockscout/docs/blob/master/for-developers/information-and-settings/env-variables/deprecated-env-variables/README.md) page. After merging docs PR, changes will be reflected in these [pages](https://docs.blockscout.com/for-developers/information-and-settings/env-variables).
- [ ] If I added new DB indices, I checked, that they are not redundant, with PGHero or other tools.
- [ ] If I added/removed chain type, I modified the Github CI matrix and PR labels accordingly.

### CODE REVIEW

File: apps/explorer/priv/repo/migrations/20240904161254_create_signed_authorizations.exs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+        primary_key: true
+      )
+
+      add(:index, :integer, null: false, primary_key: true)
+      add(:chain_id, :bigint, null: false)
+      add(:address, :bytea, null: false)
+      add(:nonce, :integer, null: false)
+      add(:v, :integer, null: false)
+      add(:r, :numeric, precision: 100, null: false)
+      add(:s, :numeric, precision: 100, null: false)
```
[the remark this answers is not in the document]
Comment (AllFi) [reply apps/explorer/priv/repo/migrations/20240904161254_create_signed_authorizations.exs:17]: Decided to keep them for now, but it makes sense to revisit this once we have a clearer understanding of how this EIP should be integrated into the Blockscout UI.

File: apps/explorer/lib/explorer/chain/import/runner/signed_authorizations.ex [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+      |> Map.take(~w(on_conflict timeout)a)
+      |> Map.put_new(:timeout, @timeout)
+      |> Map.put(:timestamps, timestamps)
+
+    Multi.run(multi, :signed_authorizations, fn repo, _ ->
+      Instrumenter.block_import_stage_runner(
+        fn -> insert(repo, changes_list, insert_options) end,
+        :block_referencing,
+        :logs,
+        :logs
```
Comment (k1rill-fedoseev) [inline apps/explorer/lib/explorer/chain/import/runner/signed_authorizations.ex:49]:
```suggestion
        :block_referencing,
        :signed_authorizations,
        :signed_authorizations
```

File: apps/ethereum_jsonrpc/lib/ethereum_jsonrpc.ex [hunk 3]
```diff
@@ -508,6 +520,21 @@ defmodule EthereumJSONRPC do
     integer
   end
 
+  @doc """
+  Converts `map()/0` to `t:signed_authorization/0`
+  """
+  @spec to_signed_authorization(map()) :: signed_authorization()
+  def to_signed_authorization(map) do
```
Comment (k1rill-fedoseev) [inline apps/ethereum_jsonrpc/lib/ethereum_jsonrpc.ex:527]: Let's move this to `apps/ethereum_jsonrpc/lib/ethereum_jsonrpc/transaction.ex`, or even better to a separate module to be consistent with how `block.withdrawals` or `receipt.logs` are handled.

**Review by k1rill-fedoseev — CHANGES_REQUESTED** [review]

File: apps/ethereum_jsonrpc/lib/ethereum_jsonrpc.ex (hunk 3 above)
Comment (AllFi) [reply apps/ethereum_jsonrpc/lib/ethereum_jsonrpc.ex]: Done in the https://github.com/blockscout/blockscout/pull/10776/commits/d3eff0dfd41cebc52ea35651a435c0297ae94b37.

**Review by vbaranov — COMMENTED** [review]
The changes look good to me in general. My only concern is that it is implied, that new functionality: an additional block import step, and preloads of signed authorisations, are spread to all chain types. Don't we want to reduce their applicability to only `ethereum` chain type?

**Comment by AllFi** [comment]
Thank you for the review!

> The changes look good to me in general. My only concern is that it is implied, that new functionality: an additional block import step, and preloads of signed authorisations, are spread to all chain types. Don't we want to reduce their applicability to only ethereum chain type?

We briefly discussed this with @k1rill-fedoseev [here](https://www.notion.so/EIP-7702-1033d73641f880f18b2feaa3d2526c0a?d=1043d73641f8807fa11c001c3cafb2fb&pvs=4#1033d73641f8806d8250ca843dd1482b) and agreed that this EIP will likely be adopted by the majority of chains. I believe the additional block import step shouldn't significantly affect performance if no signed authorizations need to be saved. Since the `signed_authorizations` table will be empty and these preloads only occur for single transaction and single address requests, I don't think they will have noticeable effect on performance as well (though I could be wrong). That’s why I believe it's better to support this for all chains from the start, to avoid the need for additional changes later.

What do you think?

**Comment by vbaranov** [comment]
> We briefly discussed this with @k1rill-fedoseev [here](https://www.notion.so/EIP-7702-1033d73641f880f18b2feaa3d2526c0a?d=1043d73641f8807fa11c001c3cafb2fb&pvs=4#1033d73641f8806d8250ca843dd1482b) and agreed that this EIP will likely be adopted by the majority of chains. I believe the additional block import step shouldn't significantly affect performance if no signed authorizations need to be saved. Since the `signed_authorizations` table will be empty and these preloads only occur for single transaction and single address requests, I don't think they will have noticeable effect on performance as well (though I could be wrong). That’s why I believe it's better to support this for all chains from the start, to avoid the need for additional changes later.
> 
> What do you think?

yep, this sounds reasonable to keep it as is then.

**Review by vbaranov — APPROVED** [review]

**Review by k1rill-fedoseev — APPROVED** [review]

**Pull request merged**